# Initial data exploration

In this notebook we will start by analysing and cleaning what is inside the `OnlineRetailSmall.csv` file. This data, taken from [here](), contains several transactions of a retail store that sells small gifts.

**Take the file from the Drive folder and upload it to this colab notebook.**

In [21]:
import pandas as pd

Let's load it up and check its size.

In [22]:
df = pd.read_csv("OnlineRetailSmall.csv")

In [23]:
df.shape

(150000, 8)

In [24]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

What are the data types?

In [25]:
df.dtypes

InvoiceNo       object
StockCode       object
Description     object
Quantity         int64
InvoiceDate     object
UnitPrice      float64
CustomerID     float64
Country         object
dtype: object

We can also get a summary of all the (numerical) columns:

In [26]:
df.describe()

Quantity      UnitPrice     CustomerID
count  150000.000000  150000.000000  115729.000000
mean        8.841127       4.288298   15327.129665
std       300.182669      72.168319    1703.006148
min    -80995.000000       0.000000   12347.000000
25%         1.000000       1.250000   14071.000000
50%         3.000000       2.080000   15186.000000
75%        10.000000       4.130000   16817.000000
max     80995.000000   17836.460000   18287.000000

# A little bit of data cleaning

How clean is our data? We can expect for example all quantities to be bigger than 0, and all unit prices to be positive, no? We can check this:

In [11]:
df[df["Quantity"] <= 0]

InvoiceNo StockCode  ... CustomerID         Country
537       570691    85231G  ...        NaN  United Kingdom
538      C570692     23169  ...    13359.0  United Kingdom
757      C570704         M  ...    13208.0  United Kingdom
762      C570708         M  ...    12908.0  United Kingdom
765       570711     21622  ...        NaN  United Kingdom
...          ...       ...  ...        ...             ...
148540   C581490     23144  ...    14397.0  United Kingdom
149632   C581499         M  ...    15498.0  United Kingdom
149806   C581568     21258  ...    15311.0  United Kingdom
149807   C581569     84978  ...    17315.0  United Kingdom
149808   C581569     20979  ...    17315.0  United Kingdom

[2368 rows x 8 columns]

Yikes! 2368 rows with negative quantities. How could we remove them? Easy:

In [12]:
positive_mask = df["Quantity"] > 0
df = df[positive_mask]

In [13]:
df

InvoiceNo StockCode  ... CustomerID         Country
0         570688     15036  ...        NaN  United Kingdom
1         570688     16236  ...        NaN  United Kingdom
2         570688    17012B  ...        NaN  United Kingdom
3         570688    17012F  ...        NaN  United Kingdom
4         570688    17084P  ...        NaN  United Kingdom
...          ...       ...  ...        ...             ...
149995    581587     22613  ...    12680.0          France
149996    581587     22899  ...    12680.0          France
149997    581587     23254  ...    12680.0          France
149998    581587     23255  ...    12680.0          France
149999    581587     22138  ...    12680.0          France

[147632 rows x 8 columns]

Now, checking the same thing for unit price:

In [15]:
df[df["UnitPrice"] <= 0]

InvoiceNo StockCode  ... CustomerID         Country
763       570709     21098  ...        NaN  United Kingdom
764       570710     21918  ...        NaN  United Kingdom
1449      570822     22501  ...        NaN  United Kingdom
1998      570860    16168M  ...        NaN  United Kingdom
2187      570868     21311  ...        NaN  United Kingdom
...          ...       ...  ...        ...             ...
143425    581211     22142  ...        NaN  United Kingdom
145072    581234     72817  ...        NaN  United Kingdom
146595    581406    46000M  ...        NaN  United Kingdom
146596    581406    46000S  ...        NaN  United Kingdom
146645    581408     85175  ...        NaN  United Kingdom

[208 rows x 8 columns]

208 rows of non-positive price. Let's remove them in a similar way:

In [16]:
df = df[df["UnitPrice"] > 0]

In [17]:
df

InvoiceNo StockCode  ... CustomerID         Country
0         570688     15036  ...        NaN  United Kingdom
1         570688     16236  ...        NaN  United Kingdom
2         570688    17012B  ...        NaN  United Kingdom
3         570688    17012F  ...        NaN  United Kingdom
4         570688    17084P  ...        NaN  United Kingdom
...          ...       ...  ...        ...             ...
149995    581587     22613  ...    12680.0          France
149996    581587     22899  ...    12680.0          France
149997    581587     23254  ...    12680.0          France
149998    581587     23255  ...    12680.0          France
149999    581587     22138  ...    12680.0          France

[147424 rows x 8 columns]

# Storing the clean dataset

We could keep cleaning our data, but let's stop here, save this version, and implement a function that summaries what we did and that could be applied to the original dataset.

In [27]:
def clean_retail_data(path: str):
  df = pd.read_csv(path)
  df = df[df["Quantity"] > 0]
  df = df[df["UnitPrice"] > 0]

  return df

In [29]:
clean_df = clean_retail_data("OnlineRetailSmall.csv")
clean_df.to_csv("OnlineRetailSmallClean.csv")